File: cobble-binding/cobble-cpp/cpp/metrics.cpp
#include <cobble/database.hpp>

#include <utility>

#include "detail/impl.hpp"

namespace cobble {

std::vector<MetricSample> Database::Metrics() const {
  if (!impl_) {
    throw Error(ErrorCode::kInvalidState, "Database has been moved from");
  }
  auto native = ffi::native_database_metrics(*impl_->native);
  std::vector<MetricSample> result;
  result.reserve(native.size());
  for (const auto& sample : native) {
    std::vector<MetricLabel> labels;
    labels.reserve(sample.labels.size());
    for (const auto& label : sample.labels) {
      labels.push_back(
          {std::string(label.key), std::string(label.value)});
    }

    MetricValue value;
    switch (sample.kind) {
      case 0:
        value = CounterValue{sample.counter};
        break;
      case 1:
        value = GaugeValue{sample.gauge};
        break;
      case 2:
        value = HistogramValue{sample.count, sample.sum, sample.min,
                               sample.max};
        break;
      default:
        throw Error(ErrorCode::kInvalidState,
                    "native metric has an unknown value kind");
    }
    result.push_back(
        {std::string(sample.name), std::move(labels), std::move(value)});
  }
  return result;
}

}  // namespace cobble

File: cobble-binding/cobble-cpp/src/metrics.rs
use crate::{database::NativeDatabase, ffi};

fn metrics(samples: Vec<cobble_binding::MetricSample>) -> Vec<ffi::NativeMetric> {
    samples
        .into_iter()
        .map(|sample| {
            let labels = sample
                .labels
                .into_iter()
                .map(|(key, value)| ffi::NativeMetricLabel { key, value })
                .collect();
            match sample.value {
                cobble_binding::MetricValue::Counter(value) => ffi::NativeMetric {
                    name: sample.name,
                    labels,
                    kind: 0,
                    counter: value,
                    gauge: 0.0,
                    count: 0,
                    sum: 0.0,
                    min: 0.0,
                    max: 0.0,
                },
                cobble_binding::MetricValue::Gauge(value) => ffi::NativeMetric {
                    name: sample.name,
                    labels,
                    kind: 1,
                    counter: 0,
                    gauge: value,
                    count: 0,
                    sum: 0.0,
                    min: 0.0,
                    max: 0.0,
                },
                cobble_binding::MetricValue::Histogram(value) => ffi::NativeMetric {
                    name: sample.name,
                    labels,
                    kind: 2,
                    counter: 0,
                    gauge: 0.0,
                    count: value.count,
                    sum: value.sum,
                    min: value.min,
                    max: value.max,
                },
            }
        })
        .collect()
}

pub(crate) fn native_database_metrics(db: &NativeDatabase) -> Vec<ffi::NativeMetric> {
    metrics(db.db.metrics())
}

pub(crate) fn native_sharded_database_metrics(
    db: &crate::sharded_db::NativeShardedDatabase,
) -> Vec<ffi::NativeMetric> {
    metrics(db.db.metrics())
}
